frontend/app/actions/RegistrationUri.scala

```diff
@@ -9,43 +9,38 @@ object RegistrationUri {
   val REFERRER_HEADER: String = "referer"
   val CAMPAIGN_SOURCE: String = "MEM"
 
+  val referingPageCodes = Map(
+    controllers.routes.Info.supporter().toString() -> "SUPUK",
+    controllers.routes.Info.supporterUSA().toString() -> "SUPUS",
+    controllers.routes.Info.offersAndCompetitions().toString()-> "COMP",
+    controllers.routes.WhatsOn.list().toString() -> "EVT",
+    "/" -> "HOME"
+  )
+
+  val campaignTierCodes = Map[String, String](
+    controllers.routes.Joiner.joinPaid(Tier.Supporter).toString() -> "SUP",
+    controllers.routes.Joiner.joinPaid(Tier.Partner).toString() -> "PAR",
+    controllers.routes.Joiner.joinPaid(Tier.Patron).toString() -> "PAT",
+    controllers.routes.Joiner.joinPaid(Tier.Friend).toString() -> "FRI"
+  )
+
   def parse(request: RequestHeader) = {
     val redirectUrl: String = Config.idWebAppRegisterUrl(request.uri)
-    extractCampaignCode(request.headers(REFERRER_HEADER), request.path) match {
-      case Some(campaignCode) => redirectUrl.concat(s"&INTCMP=${campaignCode.get}")
-      case _ => redirectUrl
-    }
+    val campaignCode = extractCampaignCode(request.headers.get(REFERRER_HEADER), request.path)
+    redirectUrl.concat(s"&INTCMP=$campaignCode")
   }
 
-  private def extractCampaignCode(referer: String, path: String) = {
+  private def extractCampaignCode(refererOpt: Option[String], path: String) : String = {
+
+    val referingCode = refererOpt.flatMap(getReferingPageCode)
+    val campaignTier: Option[String] = campaignTierCodes.get(path)
+
+    (Seq(CAMPAIGN_SOURCE) ++ referingCode ++ campaignTier).mkString("_")
+  }
+
+  def getReferingPageCode(referer: String): Option[String] = {
     val refererUrl = Uri.parse(referer)
-    val supporter = controllers.routes.Info.supporter().toString()
-    val supporterUS = controllers.routes.Info.supporterUSA().toString()
-    val competitions = controllers.routes.Info.offersAndCompetitions().toString()
-    val events = controllers.routes.WhatsOn.list().toString()
-
-    val campaignReferer = refererUrl.path match {
-      case `supporter` => "SUPUK"
-      case `supporterUS` => "SUPUS"
-      case `competitions` => "COMP"
-      case `events` => "EVT"
-      case _ => "HOME"
-    }
-
-    val supporterTier = controllers.routes.Joiner.joinPaid(Tier.Supporter).toString()
-    val partnerTier = controllers.routes.Joiner.joinPaid(Tier.Partner).toString()
-    val patronTier = controllers.routes.Joiner.joinPaid(Tier.Patron).toString()
-    val friendTier = controllers.routes.Joiner.joinPaid(Tier.Friend).toString()
-
-    val campaignTier = path match {
-      case `supporterTier` => Some("SUP")
-      case `partnerTier` => Some("PAR")
-      case `patronTier` => Some("PAT")
-      case `friendTier` => Some("FRI")
-      case _ => None
-    }
-    campaignTier.map {
-      tier => Some(Seq(CAMPAIGN_SOURCE, campaignReferer, tier).mkString("_"))
-    }
+    referingPageCodes.get(refererUrl.path)
```

Comment:
While we are at it, I would guard the `Uri.parse` as well with like `Try(Uri.parse).toOption`